### ISSUE TITLE
f32 scatter-add regression in jaxlib 0.10.2: CAS loop emitted instead of atom.add.f32, ~450x slower under index contention

### ISSUE THREAD

**Issue by vladfi1** [openxla/xla#46225]
## Summary

Starting with jaxlib **0.10.2** (still present in 0.11.0), XLA:GPU compiles a plain
f32 scatter-add to a compare-and-swap retry loop (`atom.relaxed.sys.global.cas.b32`)
instead of the native float atomic (`atom.relaxed.sys.global.add.f32`) that
jaxlib <= 0.10.1 emits. For scatters with heavily duplicated indices — the common
embedding-gradient pattern, where a whole batch may update a handful of table
rows — the CAS loop collapses: **~1 ms → ~500 ms** for a single scatter in the
repro below (RTX 3080 Ti).

This looks related to #26371 (bf16 scatter-add using CAS under contention), but
this is a fresh regression affecting **f32**, where a native atomic exists and
was used until 0.10.1.

## Repro

```python
import time
import jax
import jax.numpy as jnp
import numpy as np

def bench(f, *args, n=3):
  jax.block_until_ready(f(*args))
  t0 = time.perf_counter()
  for _ in range(n):
    out = f(*args)
  jax.block_until_ready(out)
  return (time.perf_counter() - t0) / n

N = 80 * 512  # 40960 updates
table = jnp.zeros((399, 128), jnp.float32)
upd = jnp.ones((N, 128), jnp.float32)

@jax.jit
def scat(tbl, ids, upd):
  return tbl.at[ids].add(upd)

for desc, ids_np in [
    ('uniform-random', np.random.randint(0, 399, N)),
    ('all-same-index', np.zeros(N, np.int32)),
]:
  ids = jnp.array(ids_np, jnp.int32)
  print(f'{desc}: {bench(scat, table, ids, upd)*1e3:.2f} ms')
```

## Results (RTX 3080 Ti, sm_86, driver 595.84, Linux x86_64, Python 3.13, CUDA 13 wheels)

| `jax[cuda13]` version | uniform-random | all-same-index |
|---|---|---|
| 0.9.2   | 0.4 ms | 1.1 ms |
| 0.10.0  | 0.4 ms | 1.1 ms |
| 0.10.1  | 0.4 ms | 1.2 ms |
| **0.10.2** | 0.4 ms | **507 ms** |
| **0.11.0** | 0.4 ms | **511 ms** |

## PTX evidence

jaxlib 0.10.1 (fast) — scatter kernel contains:

```
atom.relaxed.sys.global.add.f32
```

jaxlib 0.10.2 / 0.11.0 (slow) — same scatter compiles to a CAS retry loop:

```
$L__BB0_2:
        atom.relaxed.sys.global.cas.b32  %r5, [%rd6], %r29, %r22;
        setp.ne.b32  %p3, %r5, %r29;
        mov.b32      %r29, %r5;
        @%p3 bra     $L__BB0_2;
```

Native f32 atomic add handles same-address contention in hardware; the CAS loop
serializes and spins, so wall time scales with the collision count. A telltale
symptom is 100% GPU utilization with unusually low power draw (SMs busy
spinning, no real work).

## Real-world impact

In our training workload (Melee imitation learning, github.com/vladfi1/slippi-ai),
the backward pass scatters [40960, 128] gradient rows into small embedding tables
(33–13k rows) with highly repeated indices (character/action IDs — thousands of
duplicates per batch). Each such scatter went from ~1 ms to ~600 ms, making
training **10x slower end to end** (same code and data are fine on machines still
running jaxlib <= 0.10.1).

`--xla_gpu_enable_scatter_determinism_expander=true` avoids the CAS path but is
not a usable workaround here: the sort-based expansion launches ~100k tiny
kernels per training step and the workload becomes kernel-launch-bound (GPU ~80%
idle), still ~7x slower than 0.10.1.

Happy to provide full HLO dumps or run bisects against XLA nightlies if useful.

**Comment by Gattocrucco** [comment]
This should already have been fixed. See https://github.com/jax-ml/jax/issues/38806 for a workaround in the meantime.

**Comment by vladfi1** [comment]
Thanks! Should have looked more carefully at the existing issues.
